### PR TITLE
Undo selection appends genes from differential expression to user gene list

### PR DESCRIPTION
**Pull request by mweiden**
Fixes https://github.com/chanzuckerberg/cellxgene/issues/1171

## Need
When a user performs a differential expression from within a sub-selection (world) of the data and then resets the selection to all cells (universe), the differential expression results are no longer valid.

## Approach
* When the selection is reset, move the top (maxUserDefinedGenes - len(userDefinedGenes) from the differential expression results to the
list of user defined genes
* Raise maxUserDefinedGenes to 25 to give users more room and accommodate the extra genes transferred in from differential expression

### CODE REVIEW

**Comment by mweiden** [comment]
Second option for button icons:

Screen Shot 2020-02-28 at 11 17 55 PM

**Comment by colinmegill** [comment]
@mweiden HA! Excellent set of second button choices :) significant improvement on the first.

File: client/src/components/brushableHistogram/index.js [hunk 1]
```diff
@@ -523,6 +523,7 @@ class HistogramBrush extends React.PureComponent {
           </span>
           <span
             data-testclass="brushable-histogram-field-name"
+            data-testid={`${field}:field-name`}
```
Comment (bkmartinjr) [inline client/src/components/brushableHistogram/index.js]: `:field-name` is seems overly generic - we may have other fields in the future.  Maybe something more descriptive?

File: client/src/components/menubar/index.js [hunk 2]
```diff
@@ -362,30 +316,43 @@ class MenuBar extends React.Component {
...
+          <Tooltip
+            content="Subset to currently selected cells and associated metadata"
+            position="bottom"
+            hoverOpenDelay={globals.tooltipHoverOpenDelay}
           >
-            <Icon icon="double-chevron-down" />
-          </AnchorButton>
-        </Tooltip>
-        <ButtonGroup style={{ marginRight: "10px" }}>
+            <AnchorButton
```
Comment (bkmartinjr) [inline client/src/components/menubar/index.js]: suggestion - factor this out into a sub-component ans file like Clip or InformationMenu

File: client/src/reducers/controls.js [hunk 3]
```diff
@@ -74,9 +75,12 @@ const Controls = (
       };
     }
     case "reset World to eq Universe": {
+      const { userDefinedGenes, diffexpGenes } = state;
       return {
         ...state,
-        resettingInterface: false
+        resettingInterface: false,
+        userDefinedGenes: [].concat(userDefinedGenes, diffexpGenes).slice(0,globals.maxUserDefinedGenes),
```
Comment (bkmartinjr) [inline client/src/reducers/controls.js]: this removes the most recently added genes -- is that intentional?   I would have imagined we would remove the least recently added genes.

@liaprins-czi - thoughts?

**Comment by bkmartinjr** [comment]
> I think they'd still expect the genes to stay no matter what they do.

No question you are right.

@mweiden - how about changing the logic:
* you can't manually add more user defined genes when total number exceeds the global limit
* but...you can still add them via subset/reset

This leaves a loophole, where a clever user could continuously add genes, but I don't think we really care.  Worst case, they run out of memory in their browser. 

Alternatively, up the global limit to something much larger, etc, 100.

File: client/src/reducers/controls.js (hunk 3 above)
Comment (liaprins-czi) [reply client/src/reducers/controls.js]: Yes I agree it should keep the most recently added and sacrifice the "oldest" genes instead... good catch

File: client/src/components/brushableHistogram/index.js (hunk 1 above)
Comment (mweiden) [reply client/src/components/brushableHistogram/index.js]: Ah! Turns out I didn't use this anyway. Deleting it.

File: client/src/reducers/controls.js (hunk 3 above)
Comment (mweiden) [reply client/src/reducers/controls.js]: Good points. @bkmartinjr I've implemented the solution you mentioned below: don't throw out any user-added genes and have a max gene limit that only slices off diffexp genes.

**Comment by mweiden** [comment]
> - you can't manually add more user defined genes when total number exceeds the global limit
> - but...you can still add them via subset/reset
>
> Alternatively, up the global limit to something much larger, etc, 100.

**Comment by sidneymbell** [comment]
This looks awesome!

Re: number of genes, this relates to the request from multiple users to be able to specify the number of differentially expressed genes returned. Is there a strong reason to cap the number of genes so tightly? I'd be in favor of Bruce's suggestion to increase it to, say, 100. 

If the user hits the limit, yes, removing the "oldest" genes first makes sense to me.

**Comment by mweiden** [comment]
@sidneymbell responses inline below:

> Is there a strong reason to cap the number of genes so tightly? I'd be in favor of Bruce's suggestion to increase it to, say, 100.
I think you're conflating number of genes returned by differential expression with the number of genes displayed on the sidebar. I did implement a limit on the total number of genes that can be displayed on the sidebar (`globals.maxGenes`=100). Further, I increased the number of genes users can add manually to the sidebar (`globals.maxUserDefinedGenes`=25). However, the number of results returned from each run of differential expression is still 10.

> [Users request] to be able to specify the number of differentially expressed genes returned
Agree that this should be an option. Let's follow up on this as part of the sidebar refactor.

> If the user hits the limit, yes, removing the "oldest" genes first makes sense to me.
With the changes in 7e2ce04, no genes manually added by the user will be removed.

**Pull request merged**